### PR TITLE
Update spec + packages to split functions for links from linkedAggregations

### PR DESCRIPTION
**Pull request by CiaranMn**
We have the concept of two types of 'links' in the protocol:
1. A single target link, which is a traditional link between an entity and another entity, i.e. an edge in a graph
2. A 'linked aggregation', which is a link from a single entity to an 'aggregation' of entities, e.g. some collection of entities which represents the result of some operation/filter.

These were previously defined using the same type and managed using the same operations, which was unwieldy.

This PR splits them out into separate definitions, managed by separate functions. It:
- updates the types in `blockprotocol`
- updates the spec
- updates `mock-block-dock`
- bumps the versions of HASH blocks in the hub

## To test

1. Try out different blocks in the Block Hub.
2. Read the updated spec page for linking entities at `/spec/block-types#linking-entities` and check it makes sense in explaining the two types (links, linkedAggregations)

## Next steps

Update HASH to remove the patch changes introduced in https://github.com/hashintel/hash/pull/497, which should no longer required (a few minor differences to the types here, notably renaming `LinkedAggregationInput` to `LinkedAggregationDefinition`)

### CODE REVIEW

File: site/src/_pages/spec/1_block-types.mdx [hunk 1]
```diff
@@ -204,12 +206,14 @@ Any operations specified in the protocol should be useful in a wide variety of c
 </Frame>
 
 ```block-function
-createEntities<T>(actions: CreateEntitiesAction<T>[]): Promise<T[]>
+createEntities(actions: CreateEntitiesAction[]): Promise<Entity[]>
```
Comment (CiaranMn) [inline site/src/_pages/spec/1_block-types.mdx:209]: These generics were confusing some people and aren't accurate anyway, since the returned thing has more properties than the input (e.g. identifying fields)

File: packages/mock-block-dock/src/util.ts [hunk 2]
```diff
@@ -106,7 +127,11 @@ const sortEntitiesOrTypes = <
 
   return orderBy(
     [...entities],
-    multiSort.map(({ field }) => field),
+    multiSort.map(
+      ({ field }) =>
+        (entity) =>
+          get(entity, field),
+    ),
```
Comment (CiaranMn) [inline packages/mock-block-dock/src/util.ts:134]: Drive-by: fix sorting on nested properties.

File: packages/block-template/package.json [hunk 3]
```diff
@@ -24,11 +24,11 @@
     "serve": "block-scripts serve"
```
Comment (kachkaev) [inline packages/block-template/package.json:24]: I guess we need a new version for `block-template` given the dependency changes. As a separate task, we can check feasibility of using dependency ranges (`^`) to release downstream packages less frequently.

**Review by kachkaev — COMMENTED** [review]
The only thing I’ve spotted is bumping `block-template` version. @benwerner01 WDYT?

File: packages/mock-block-dock/src/useMockDatastore.ts [hunk 4]
```diff
@@ -176,7 +180,92 @@ export const useMockDatastore = (
...
+              matchEntityTypeIdentifiers(entityType, action),
+            );
+            if (isMatch) {
+              return null;
+            }
+            return entityType;
+          })
+          .filter(
+            (entityType): entityType is BlockProtocolEntityType => !!entityType,
+          ),
```
Comment (benwerner01) [inline packages/mock-block-dock/src/useMockDatastore.ts:264]: Not sure if I'm missing something obvious, but why not just:
```suggestion
        currentEntityTypes
          .filter(
            (entityType) => !actions.some((action) =>
              matchEntityTypeIdentifiers(entityType, action),
            ),
          ),
```
So that you can implement the boolean return array more easily down the line?

**Review by benwerner01 — CHANGES_REQUESTED** [review]
Thanks Ciaran, I think I've spotted some blocking things (some of which I'm less sure about than others).

The remainder are minor suggestions/thoughts - feel free to ignore any you disagree with.

File: packages/blockprotocol/core.d.ts [hunk 5]
```diff
@@ -295,43 +276,102 @@ export type BlockProtocolGetLinksFunction = {
...
+export type BlockProtocolGetLinkedAggregationsFunction = {
+  (actions: BlockProtocolGetLinkedAggregationAction[]): Promise<
+    BlockProtocolLinkedAggregation[]
+  >;
+};
+
+export type BlockProtocolCreateLinkedAggregationAction = Omit<
+  BlockProtocolLinkedAggregationDefinition,
+  "aggregationId"
+>;
```
[the remark this answers is not in the document]
Comment (CiaranMn) [reply packages/blockprotocol/core.d.ts:329]: https://github.com/blockprotocol/blockprotocol/pull/302#discussion_r865967376

File: packages/blockprotocol/core.d.ts [hunk 6]
```diff
@@ -295,43 +276,102 @@ export type BlockProtocolGetLinksFunction = {
   (actions: BlockProtocolGetLinkAction[]): Promise<BlockProtocolLink[]>;
 };
 
-export type BlockProtocolCreateLinksAction = DistributedOmit<
-  BlockProtocolLink,
-  "linkId"
->;
+export type BlockProtocolCreateLinksAction = Omit<BlockProtocolLink, "linkId">;
```
[the remark this answers is not in the document]
Comment (CiaranMn) [reply packages/blockprotocol/core.d.ts:273]: https://github.com/blockprotocol/blockprotocol/pull/302#discussion_r865967376

File: packages/blockprotocol/core.d.ts [hunk 7]
```diff
@@ -295,43 +276,102 @@ export type BlockProtocolGetLinksFunction = {
...
+
+export type BlockProtocolUpdateLinkedAggregationActionFragment = {
+  sourceAccountId?: string | null;
+  aggregationId: string;
+};
+
+export type BlockProtocolUpdateLinkedAggregationAction = {
+  sourceAccountId?: string | null;
+  aggregationId: string;
+  data: BlockProtocolLinkedAggregationDefinition["operation"];
```
[the remark this answers is not in the document]
Comment (CiaranMn) [reply packages/blockprotocol/core.d.ts]: Yes - updated [here](https://github.com/blockprotocol/blockprotocol/pull/302/commits/c68c4bf9bbd327ab5559d93b7dec9937bb1af28c), and will need to merge [this HASH PR](https://github.com/hashintel/hash/pull/533) and update the table block commit in this one once it's merged

File: site/src/_pages/spec/1_block-types.mdx [hunk 8]
```diff
@@ -549,6 +523,102 @@ retrieve one or more links.
...
+To do so, blocks SHOULD create a `LinkedAggregationDefinition`, which:
+
+- MUST contain:
+
+  - `sourceEntityId` \[_string_]: the `entityId` of the source entity.
+  - `operation` – an aggregation operation which the embedding application should resolve the link to, following the structure of the `operation` object [described above](#aggregate-entities-operation).
+  - `path` \[_string_]: the path to the field on the source entity this link is conceptually made on, expressed as a [JSON path](https://goessner.net/articles/JsonPath/).
+
+- MAY contain
+  - `sourceEntityVersionId?` \[_string_]\[_optional_]: optionally specify that this link is only from a specific version.
```
[the remark this answers is not in the document]
Comment (benwerner01) [reply site/src/_pages/spec/1_block-types.mdx]: Thank you Ciaran, that makes sense 👍

File: site/src/_pages/spec/1_block-types.mdx [hunk 9]
```diff
@@ -112,9 +112,9 @@ Blocks SHOULD expect the following additional fields related to linked data to b
 
 - a `linkGroups` field which provides the links attached to the block or the entities provided in `linkedEntities`, grouped by entity and path, i.e. these are the **edges** of the graph.
 
-- a `linkedAggregations` field containing the results of aggregations linked from the block’s entity - these are special links which represent an aggregation operation (e.g. in plain English, "top 10 Cars sorted by age, descending"), which are provided in this field along with the results of the operation at the time of request.
+- a `linkedAggregations` field containing the definitions and results of aggregations linked from the block’s entity - `LinkedAggregations` are special objects which contain an aggregation definition (e.g. in plain English, "top 10 Cars sorted by age, descending") and its results at the time of provision.
```
[the remark this answers is not in the document]
Comment (CiaranMn) [reply site/src/_pages/spec/1_block-types.mdx:115]: I've made this consistent (by removing the `BlockProtocolEntity` for now - I started prefixing everything with `BlockProtocol`, but this applies to all the function names and signatures too, and it felt like it was starting to get in the way of understanding. Maybe we should remove it from the TS types too. But I would like to leave that decision out of scope of this PR. https://github.com/blockprotocol/blockprotocol/pull/302/commits/618ce2a25736a0f8a9c38acaab9bf6fdac2486d5

File: packages/mock-block-dock/src/useMockDatastore.ts (hunk 4 above)
Comment (CiaranMn) [reply packages/mock-block-dock/src/useMockDatastore.ts]: I think I just copied this code from other methods and didn't think about it properly.

I've done the boolean return properly [here](https://github.com/blockprotocol/blockprotocol/pull/302/files/618ce2a25736a0f8a9c38acaab9bf6fdac2486d5..c2a41e205081a3630491aae8f5720d9b2edd5323)

File: packages/blockprotocol/core.d.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (benwerner01) [reply packages/blockprotocol/core.d.ts]: And I'll need to update my chart block :)
